### ISSUE TITLE
`self-assigning-variable` - option to allow reassignment with type annotation

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#28688]
### Summary

some codebases allow reassignment to a variable if it has a type annotation ie. shadowing:

```py
foo = 1

def fn(foo: int):
    ...
    foo: int | str = foo # error: self-assigning-variable
```
it would be nice if ruff had an option to allow this, especially since it's supported by most type checkers including ty. (in pyright it's configurable using `reportRedeclaration`)

**Comment by ntBre** [comment]
This already sounded reasonable to me as a configuration option, but apparently this is just allowed by pylint:

```pycon
❯ uvx pylint==4.0.8 --rcfile=/dev/null \
  --disable=all --enable=self-assigning-variable \
  --reports=no --score=no --persistent=no \
  --from-stdin example.py <<'PY'
def annotated(foo: int):
    foo: int | str = foo

def plain(foo: int):
    foo = foo
PY
************* Module example
example.py:5:4: W0127: Assigning the same variable 'foo' to itself (self-assigning-variable)
```

I might lean toward just changing the rule's default behavior and considering this more of a bug.

**Comment by DetachHead** [comment]
That sounds good too, and maybe a separate rule to ban redeclarations
